src/wasm-traversal.h

```diff
@@ -157,185 +149,209 @@ struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
-    ExpressionList& list = curr->operands;
-    for (size_t z = 0; z < list.size(); z++) {
-      parent.walk(list[z]);
-    }
+  struct Task {
+    TaskFunc func;
+    Expression** currp;
+    Task(TaskFunc func, Expression** currp) : func(func), currp(currp) {}
+  };
+
+  std::vector<Task> stack;
+
+  void addTask(TaskFunc func, Expression** currp) {
+    stack.emplace_back(func, currp);
   }
-  void visitCallIndirect(CallIndirect *curr) {
-    parent.walk(curr->target);
-    ExpressionList& list = curr->operands;
-    for (size_t z = 0; z < list.size(); z++) {
-      parent.walk(list[z]);
+  void maybeAddTask(TaskFunc func, Expression** currp) {
+    if (*currp) {
+      stack.emplace_back(func, currp);
     }
   }
-  void visitGetLocal(GetLocal *curr) {}
-  void visitSetLocal(SetLocal *curr) {
-    parent.walk(curr->value);
-  }
-  void visitLoad(Load *curr) {
-    parent.walk(curr->ptr);
-  }
-  void visitStore(Store *curr) {
-    parent.walk(curr->ptr);
-    parent.walk(curr->value);
-  }
-  void visitConst(Const *curr) {}
-  void visitUnary(Unary *curr) {
-    parent.walk(curr->value);
-  }
-  void visitBinary(Binary *curr) {
-    parent.walk(curr->left);
-    parent.walk(curr->right);
-  }
-  void visitSelect(Select *curr) {
-    parent.walk(curr->ifTrue);
-    parent.walk(curr->ifFalse);
-    parent.walk(curr->condition);
-  }
-  void visitReturn(Return *curr) {
-    parent.walk(curr->value);
-  }
-  void visitHost(Host *curr) {
-    ExpressionList& list = curr->operands;
-    for (size_t z = 0; z < list.size(); z++) {
-      parent.walk(list[z]);
+  Task getTask() {
+    auto ret = stack.back();
+    stack.pop_back();
+    return ret;
+  }
+
+  void walk(Expression*& root) {
+    assert(stack.size() == 0);
+    addTask(SubType::scan, &root);
+    while (stack.size() > 0) {
+      auto task = getTask();
+      assert(*task.currp);
+      task.func(static_cast<SubType*>(this), task.currp);
+      if (replace) {
+        *task.currp = replace;
+        replace = nullptr;
+      }
     }
   }
-  void visitNop(Nop *curr) {}
-  void visitUnreachable(Unreachable *curr) {}
-};
-
-// Walker that allows replacements
-template<typename SubType, typename ReturnType = void>
-struct WasmReplacerWalker : public WasmWalkerBase<SubType, ReturnType> {
-  Expression* replace = nullptr;
-
-  // methods can call this to replace the current node
-  void replaceCurrent(Expression *expression) {
-    replace = expression;
-  }
 
-  void walk(Expression*& curr) override {
-    if (!curr) return;
-
-    this->visit(curr);
-
-    if (replace) {
-      curr = replace;
-      replace = nullptr;
-    }
-  }
+  // subclasses implement this to define the proper order of execution
+  static void scan(SubType* self, Expression** currp) { abort(); }
+
+  // task hooks to call visitors
+
+  static void doVisitBlock(SubType* self, Expression** currp)        { self->visitExpression(*currp); self->visitBlock((*currp)->cast<Block>()); }
```

Comment:
I tried some more now to get something like this to work. I can't quite seem to find a good way that doesn't have other downsides (like relying on overloading of `visit(Expression*)` vs `visit(Block*)`, which seems potentially confusing if people forget to downcast), and that compiles ;) Perhaps if we find a good approach here we can do it as a followup.